cmd/consul-dataplane/config_test.go

```diff
@@ -0,0 +1,547 @@
+								BearerToken:     "bearer-login",
+								BearerTokenPath: "/consul/bearertokenpath/",
+								Namespace:       "default",
+								Partition:       "default",
+								Datacenter:      "dc100",
+							},
+						},
+						TLS: &consuldp.TLSConfig{
+							Disabled:           false,
+							CACertsPath:        "/consul/",
+							CertFile:           "ca-cert.pem",
+							KeyFile:            "key.pem",
+							ServerName:         "tls-server-name",
+							InsecureSkipVerify: true,
+						},
+					},
+					Service: &consuldp.ServiceConfig{
+						NodeName:  "test-node-dc1",
+						NodeID:    "dc1.node.id",
+						Namespace: "default",
+						ServiceID: "node1.service1",
+						Partition: "default",
+					},
+					Logging: &consuldp.LoggingConfig{
+						Name:     DefaultLogName,
+						LogJSON:  false,
+						LogLevel: "WARN",
+					},
+					DNSServer: &consuldp.DNSServerConfig{
+						BindAddr: "127.0.0.2",
+						Port:     8604,
+					},
+					XDSServer: &consuldp.XDSServer{
+						BindAddress: "127.0.1.0",
+						BindPort:    6060,
+					},
+					Envoy: &consuldp.EnvoyConfig{
+						AdminBindAddress:              "127.0.1.0",
+						AdminBindPort:                 18000,
+						ReadyBindAddress:              "127.0.1.0",
+						ReadyBindPort:                 18003,
+						EnvoyConcurrency:              4,
+						EnvoyDrainStrategy:            "test-strategy",
+						ShutdownDrainListenersEnabled: true,
+						GracefulShutdownPath:          "/graceful_shutdown",
+						EnvoyDrainTimeSeconds:         30,
+						GracefulPort:                  20300,
+						DumpEnvoyConfigOnExitEnabled:  true,
+					},
+					Telemetry: &consuldp.TelemetryConfig{
+						UseCentralConfig: true,
+						Prometheus: consuldp.PrometheusTelemetryConfig{
+							RetentionTime: 10 * time.Second,
+							ScrapePath:    "/metrics",
+							MergePort:     12000,
+							CACertsPath:   "/consul/",
+							CertFile:      "prom-ca-cert.pem",
+							KeyFile:       "prom-key.pem",
+						},
+					},
+				}
+
+				return reflect.DeepEqual(cfg, expectedCfg)
+			},
+			wantErr: false,
+		},
+		{
+			desc: "able to generate config properly when config file is given without flag inputs",
+			flagOpts: func() (*FlagOpts, error) {
+				opts := &FlagOpts{}
+				opts.configFile = "test.json"
+				return opts, nil
+			},
+			writeConfigFile: func() error {
+				inputJson := `{
+					"consul": {
+					  "addresses": "consul_server.dc1",
+					  "grpcPort": 8502,
+					  "serverWatchDisabled": false
+					},
+					"service": {
+					  "nodeName": "test-node-1",
+					  "serviceId": "frontend-service-sidecar-proxy",
+					  "namespace": "default",
+					  "partition": "default"
+					},
+					"envoy": {
+					  "adminBindAddress": "127.0.0.1",
+					  "adminBindPort": 19000
+					},
+					"logging": {
+					  "logLevel": "info",
+					  "logJSON": false
+					}
+				  }`
+
+				err := os.WriteFile("test.json", []byte(inputJson), 0600)
+				if err != nil {
+					return err
+				}
```

Comment:
minor: Have this function accept a `*testing.T` so that we can run assertions here and use `t.Cleanup` to run cleanups after the test finishes.

```go
			writeConfigFile: func(t *testing.T) {
```

And then this becomes:

```suggestion
				err := os.WriteFile("test.json", []byte(inputJson), 0600)
				require.NoError(t, err)
				t.Cleanup(func() { _ = os.Remove("test.json") }
```

I prefer this because the cleanups are added right next to where the file is written. (And can make it easier to refactor, if we want to move this function somewhere else because the cleanups are self-contained). What do you think?